lark/parser_frontends.py

```diff
@@ -27,56 +26,112 @@ def lex(self, lexer_state, parser_state):
+        lexer_conf.use_bytes = options.use_bytes
+        lexer_conf.g_regex_flags = options.g_regex_flags
+        lexer_conf.skip_validation = True
+        lexer_conf.postlex = options.postlex
+
+        return ParsingFrontend(lexer_conf, parser_conf, options, parser=parser)
+
+
+
+
+class ParsingFrontend(Serialize):
+    __serialize_fields__ = 'lexer_conf', 'parser_conf', 'parser', 'options'
+
+    def __init__(self, lexer_conf, parser_conf, options, parser=None):
+        self.parser_conf = parser_conf
+        self.lexer_conf = lexer_conf
+        self.options = options
+
+        # Set-up parser
+        if parser:  # From cache
+            self.parser = parser
+        else:
+            create_parser = {
+                'lalr': create_lalr_parser,
+                'earley': make_early,
+                'cyk': CYK_FrontEnd,
+            }[parser_conf.name]
+            self.parser = create_parser(lexer_conf, parser_conf, options)
+
+        # Set-up lexer
+        self.skip_lexer = False
+        if lexer_conf.name in ('dynamic', 'dynamic_complete'):
+            self.skip_lexer = True
+            return
+
+        try:
+            create_lexer = {
+                'standard': create_traditional_lexer,
+                'contextual': create_contextual_lexer,
+            }[lexer_conf.name]
+        except KeyError:
+            assert issubclass(lexer_conf.name, Lexer), lexer_conf.name
+            self.lexer = _wrap_lexer(lexer_conf.name)(lexer_conf)
+        else:
+            self.lexer = create_lexer(lexer_conf, self.parser, lexer_conf.postlex)
+
+        if lexer_conf.postlex:
+            self.lexer = PostLexConnector(self.lexer, lexer_conf.postlex)
+
+
+    def _parse(self, start, input, *args):
+        if start is None:
+            start = self.parser_conf.start
+            if len(start) > 1:
+                raise ConfigurationError("Lark initialized with more than 1 possible start rule. Must specify which start rule to parse", start)
+            start ,= start
+        return self.parser.parse(input, start, *args)
+
+    def parse(self, text, start=None):
+        if self.skip_lexer:
+            return self._parse(start, text)
+
+        lexer = LexerThread(self.lexer, text)
+        return self._parse(start, lexer)
+
+
 def get_frontend(parser, lexer):
     if parser=='lalr':
         if lexer is None:
             raise ConfigurationError('The LALR parser requires use of a lexer')
-        elif lexer == 'standard':
-            return LALR_TraditionalLexer
-        elif lexer == 'contextual':
-            return LALR_ContextualLexer
-        elif issubclass(lexer, Lexer):
-            wrapped = _wrap_lexer(lexer)
-            class LALR_CustomLexerWrapper(LALR_WithLexer):
-                def init_lexer(self):
-                    self.lexer = wrapped(self.lexer_conf)
-            return LALR_CustomLexerWrapper
-        else:
+        if lexer not in ('standard' ,'contextual') and not issubclass(lexer, Lexer):
```

Comment:
If `dynamic` gets here in `lexer`, this will throw TypeError. I would suggest splitting this up into multiple if statements, or using `isinstance(lexer, type)`. In theory, `LexerOptions` already checks if it is a subclass.

Reply:
Done